Add Typescript Types

miles-grant-ibigroup opened the pull request:
Have added types to the main file as well as a `tsconfig.json` file. Tests are not working at the moment due to working out mastarm issues.

miles-grant-ibigroup committed:
refactor(index.ts): add Typescript types

miles-grant-ibigroup committed:
refactor(index.ts): improve typescript style

miles-grant-ibigroup committed:
refactor(package.json): update main to reflect new file extension

miles-grant-ibigroup committed:
refactor(index.ts): export types

this causes a combination of commonjs and es6

miles-grant-ibigroup committed:
chore: remove d.ts file not needed due to new typing

miles-grant-ibigroup committed:
style: format files based on new mastarm linting

miles-grant-ibigroup committed:
refactor(index.ts): mastarm linting, fixes based on typing

miles-grant-ibigroup committed:
test: update tests to work in typescript environment

miles-grant-ibigroup committed:
ci: move from travis to github actions

miles-grant-ibigroup committed:
refactor: typos and small corrections to ensure tests run correctly

trevorgerhardt commented:
Hi @miles-grant-ibigroup, thanks for contributing. I have a few thoughts/notes on this migration to TypeScript. 

1. `"main": "index.js"`
The main file exported **needs** to be transpiled to JavaScript. 

2. [TSDX](https://tsdx.io/)
I would prefer to adopt TSDX. It's a "zero-config" TypeScript linter, test runner, formatter, and transpiler aimed at packages (as opposed to applications). It should work out of the box for this package and allow us to even remove configuration files keeping everything simple. A new real world example of how it's used can be seen here: https://github.com/placemark/check-geojson/blob/main/package.json

3. [@types/geojson](https://www.npmjs.com/package/@types/geojson)
If possible, let's specifically include `GeoJSON.Point` as an allowable input. It may overlap with how we define `Point` but the type hints around it are useful.

Additionally, we should consider swapping the internal names of `Coordinate` and `Point` to better align with `GeoJSON`.

miles-grant-ibigroup committed:
refactor(index.ts): add es6 exports, fix some types

miles-grant-ibigroup committed:
refactor(index.ts): use GeoJSON point and position types

evansiroky committed:
chore: use tsdx instead of mastarm, update documentation pkg

miles-grant-ibigroup committed:
Merge branch 'tsdx' of github.com:evansiroky/lonlat

miles-grant-ibigroup committed:
chore(package.json): fix coverage reporting

jest cover by default isn't smart enough to map to index.ts, so use the built version

miles-grant-ibigroup committed:
refactor: move to es6 imports and exports

miles-grant-ibigroup committed:
refactor: fix small type errors discovered by moving tests to typescript

miles-grant-ibigroup commented:
Hi Trevor,

Thanks so much for your suggestions! After some discussion we've implemented all of these. 

As to swapping `Coordinate` and `Point`, this would cause breaking changes (`fromCoordinate` and `fromPoint` are backwards, right?) so I think we might hold off on that as part of this migration, unless you have a strong desire not to!

Please let me know if there's anything else you need from me.

evansiroky commented on .github/workflows/node-ci.yml:
```diff
@@ -0,0 +1,26 @@
+          node-version: ${{ matrix.node-version }}
+      - name: Install npm packages using cache
+        uses: bahmutov/npm-install@v1
+      - name: Run Tests
+        run: npm run test
```
Please also add a step for linting.

evansiroky replied:
There also needs to be a step for semantic-release.

evansiroky commented on package.json:
```diff
@@ -31,10 +34,25 @@
+    "@types/node": "^15.12.5",
+    "documentation": "^13.2.5",
+    "eslint": "^7.30.0",
     "semantic-release": "^15.10.3",
```
Please update to the latest.

evansiroky replied:
Hasn't been updated...

miles-grant-ibigroup replied:
Should've wrote what happened here sorry -- the actual version in the yarn.lock is the newest, but I think since it's still compatible with the version here, the package.json hasn't changed. We could explicitly update the version here if you think that's a good idea?

miles-grant-ibigroup replied:
With `semantic-release` it's the same thing. The version is updated in the lock file, but `yarn upgrade` doesn't update the package.json because the latest version is still compatible. Should we explicitly update?

miles-grant-ibigroup replied:
Was doing the upgrades incorrectly, sorry. Should be fixed now!

evansiroky commented on package.json:
```diff
@@ -31,10 +34,25 @@
+    "eslint": "^7.30.0",
     "semantic-release": "^15.10.3",
-    "travis-deploy-once": "^5.0.9"
+    "travis-deploy-once": "^5.0.9",
```
This will probably not be needed anymore.

evansiroky reviewed (changes requested):
@miles-grant-ibigroup Please add the relevant items to the GitHub Actions script and make the requested package updates as well. This should be good to go pending @trevorgerhardt's approval after that. I'm also tagging @landonreed so he sees what we've done (I can't assign people in this repo).

In our work together, @miles-grant-ibigroup and I think we will probably fork TSDX in order to be able to better implement some custom eslint rules we'd like to use. Surprisingly, TSDX has already effectively stopped being maintained as the original maintainer is now working on a paid version that's supposed to work well with monorepos. 

While it's kind of nice to have jest, lint and transpiling all coming with the same package, any customizations are somewhat difficult. eslint, jest in particular seem to do relatively well on a standalone basis, so perhaps there might be less pain in trying to customize a package that does all 3 of those things than to individually maintain small direct customizations for each within each repo. That also might have an added benefit of not having to change so many things all at once. But I suppose it does run the risk of changes to one of the items breaking the other. Regardless, this TSDX approach here isn't quite perfect, but will probably work well enough for now and we concluded it was probably much less effort than trying to refactor mastarm to do all of this.

miles-grant-ibigroup committed:
refactor: address pr suggestions

miles-grant-ibigroup commented:
Have made all those fixes thanks! I believe the one last commit to make here is to move to our forked TSDX? Please correct me if I'm wrong.

Finally, I think the CI build is still failing because it's set to use Travis instead of GitHub Actions. I think that will be something @trevorgerhardt has to change.

trevorgerhardt commented on .github/workflows/node-ci.yml:
```diff
@@ -0,0 +1,37 @@
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
```
Tests should be run on each `push` and `pull_request`, not just on master.

Only semantic release should be restricted to being run on master.

trevorgerhardt commented on tsconfig.json:
```diff
@@ -0,0 +1,73 @@
+
+    /* Experimental Options */
+    // "experimentalDecorators": true,              /* Enables experimental support for ES7 decorators. */
+    // "emitDecoratorMetadata": true,               /* Enables experimental support for emitting type metadata for decorators. */
```
Any particular reason we're leaving all the commented out options inline?

miles-grant-ibigroup replied:
This is leftover from initializing the project as a Typescript project. The entire file can be removed now that things are moved to tsdx!

trevorgerhardt reviewed (commented):
Minor nitpicking note on `export` style.

My preference would be:
```js
export function functionName () {}
```
Over
```js
const functionName = () => {}
export { functionName }
```

But I don't think this should hold up the changes. Other comments I have made inline.

evansiroky commented on .gitignore:
```diff
@@ -32,3 +33,9 @@ node_modules
+.DS_Store
+
+# Typescript type files
+*.d.ts
```
We'll need to add a .npmignore file to make sure these type definition files don't get ignored when preparing a release. See [docs](https://docs.npmjs.com/cli/v7/using-npm/developers#keeping-files-out-of-your-package).

evansiroky reviewed (changes requested):
- Please update semantic-release to the latest version (see https://github.com/conveyal/lonlat/pull/19#discussion_r668486515)
- Please create a .npmignore file to make sure type definitions get included.

miles-grant-ibigroup committed:
refactor: update tsdx to ibi fork

miles-grant-ibigroup committed:
chore: add .npmignore so type definitions are published

miles-grant-ibigroup committed:
refactor: address pr feedback

miles-grant-ibigroup commented:
Thanks so much for all the polishing suggestions. The weird export style is a relic from when we were figuring out the best way to do the transpiring and still had the cjs exports. Made more sense to have them both there in the same style. But now that those are gone you're right it makes more sense to do it like this!

miles-grant-ibigroup committed:
chore: fix dependency issues from forked tsdx

miles-grant-ibigroup committed:
ci: fix dependency issues to allow ci to install

miles-grant-ibigroup committed:
chore: update dependencies

miles-grant-ibigroup commented:
Am now working on our tsdx fork. Looks like it has some problems which cause the ci to fail to install packages. The solution is to disable yarn network concurrency, which I've added here for now.

But please let me know if there are any other issues within lonlat to address!

miles-grant-ibigroup committed:
refactor: update tsdx version

miles-grant-ibigroup committed:
ci: disable network concurrency to install tsdx

evansiroky commented on package.json:
```diff
@@ -1,17 +1,21 @@
     "semantic-release": "semantic-release",
-    "test": "yarn run lint && yarn run lint-docs && yarn run jest",
-    "travis-deploy-once": "travis-deploy-once"
+    "test": "yarn run jest && yarn run lint-docs",
```
The previous `test` command should still be used here to make sure linting is included.

evansiroky reviewed (changes requested):
Please change`test` script and add the `yarn.lock` file to .npmignore. Everything else looks good to me!

Also, we need to make sure there is at least a `fix` or `feat` at the start of a commit message in this PR in order to trigger a new release. So when you make an additional commit be sure to use `fix` or `feat`.

evansiroky commented on .npmignore:
```diff
@@ -0,0 +1,3 @@
+# Coverage directory used by tools like istanbul
+coverage
+tmp
```
Please also add `yarn.lock` in here.

miles-grant-ibigroup committed:
fix: move lint to test step

to maintain backwards compatibility

miles-grant-ibigroup commented:
Thanks so much Evan! Have implemented this.

evansiroky commented on .github/workflows/node-ci.yml:
```diff
@@ -0,0 +1,34 @@
+        uses: bahmutov/npm-install@v1
+        with:
+          install-command: yarn --frozen-lockfile --network-concurrency 1
+      - name: Run Lint and Tests
+        run: yarn test
```
It's kind of nice to break these out into individual commands so that it is clear which aspect of linting/testing failed in the CI build report.

miles-grant-ibigroup replied:
Won't it lint twice then?

miles-grant-ibigroup committed:
ci: seperate out ci stages

miles-grant-ibigroup committed:
docs(ci): explanation for yarn concurrency disabling

evansiroky reviewed (approved):
OK, this looks pretty good to me. What do you say @landonreed @binh-dam-ibigroup @trevorgerhardt? Is this mergeable?

trevorgerhardt commented:
Looks good overall. Any idea why the tests aren't running?

miles-grant-ibigroup commented:
I think the `conveyal/lonlat` repo is still using travis to run the tests but the travis config was removed and replaced with a GitHub Actions config.x

binh-dam-ibigroup commented on .github/workflows/node-ci.yml:
```diff
@@ -0,0 +1,41 @@
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x]
```
This can be removed, and the node version can be inlined in the `with` > `node-version` below, e.g. [see here](https://github.com/opentripplanner/otp-react-redux/blob/dev/.github/workflows/node-ci.yml#L19)

miles-grant-ibigroup replied:
Good catch thank you! Fixed in 1711f39

binh-dam-ibigroup commented on index.ts:
```diff
@@ -112,9 +146,17 @@ module.exports.fromCoordinates = module.exports.fromGeoJSON = fromCoordinates
  */
-module.exports.fromLatlng = module.exports.fromLeaflet = function fromLatlng (lonlat) {
+export const fromLatlng = function (
+  lonlat:
+    | {
+        longitude: number
+        latitude: number
+      }
+    | { lat: number; lng: number }
```
This type could be extracted.

miles-grant-ibigroup replied:
All done in 1711f39 thanks!

binh-dam-ibigroup commented on index.ts:
```diff
@@ -129,10 +171,9 @@ module.exports.fromLatlng = module.exports.fromLeaflet = function fromLatlng (lo
  */
-function fromPoint (point) {
-  return floatize({lon: point.x, lat: point.y})
+export const fromPoint = function (point: Point): LonLatOutput {
```
I personally prefer this syntax for functions (also for all funcs in this file).

```suggestion
export function fromPoint (point: Point): LonLatOutput {
```

trevorgerhardt replied:
This is my preferred style also.

miles-grant-ibigroup replied:
I did this weird double one so that we could have the space between the function name and the parameters (since prettier squishes them otherwise) but I do agree shorter is better. Changed in 1711f39

binh-dam-ibigroup reviewed (commented):
I am not really a fan of the long declaration of exported functions `export const toString = function (input: LonLatInput): string {...` but other than that it looks good, see some minor comments.

trevorgerhardt commented:
> I think the `conveyal/lonlat` repo is still using travis to run the tests but the travis config was removed and replaced with a GitHub Actions config.x

Yes, I guess since it's a fork then Github Actions will not run locally until it's merged.

miles-grant-ibigroup committed:
refactor: address pr comments

miles-grant-ibigroup committed:
ci: correctly set node version

miles-grant-ibigroup commented:
Ah that makes sense @trevorgerhardt 
@binh-dam-ibigroup I've made the suggested changes

landonreed commented on index.ts:
```diff
@@ -419,32 +470,45 @@ module.exports.pixelToLatitude = pixelToLatitude
+    )
   }
-  return {lon: lon, lat: lat}
+  return { lat: lat, lon: lon }
```
Nit: is there a reason this line doesn't use the simpler syntax?
```suggestion
  return { lat, lon }
```

miles-grant-ibigroup replied:
Good catch thank you! I think this may be a relic from before es6.

landonreed reviewed (approved):
This looks great. Jumping in at the 11th hour, but everything y'all have done makes sense to me (just one very minor syntax nit).

miles-grant-ibigroup committed:
style: use es6 object shorthand

miles-grant-ibigroup committed:
ci: fully remove version matrix

trevorgerhardt reviewed (approved):
LGTM. Further changes can be made in other PRs.

evansiroky commented:
> LGTM. Further changes can be made in other PRs.

Thanks @trevorgerhardt! I think you might have to merge this PR since I don't think any of us have write access to this repository.
